internal/jimm/group/group.go

```diff
@@ -0,0 +1,146 @@
+	"github.com/canonical/jimm/v3/internal/dbmodel"
+	"github.com/canonical/jimm/v3/internal/errors"
+	"github.com/canonical/jimm/v3/internal/openfga"
+)
+
+// permissionManager provides a means to manage roles within JIMM.
+type groupManager struct {
+	store   *db.Database
+	authSvc *openfga.OFGAClient
+}
+
+// NewGroupManager returns a new group manager that provides group
+// creation, modification, and removal.
+func NewGroupManager(store *db.Database, authSvc *openfga.OFGAClient) (*groupManager, error) {
+	if store == nil {
+		return nil, errors.E("role store cannot be nil")
+	}
+	if authSvc == nil {
+		return nil, errors.E("role authorisation service cannot be nil")
+	}
+	return &groupManager{store, authSvc}, nil
+}
+
+// AddGroup creates a group within JIMMs DB for reference by OpenFGA.
+func (j *groupManager) AddGroup(ctx context.Context, user *openfga.User, name string) (*dbmodel.GroupEntry, error) {
+	const op = errors.Op("jimm.GetGroupManager().AddGroup")
+
+	if !user.JimmAdmin {
+		return nil, errors.E(op, errors.CodeUnauthorized, "unauthorized")
+	}
+
+	ge, err := j.store.AddGroup(ctx, name)
+	if err != nil {
+		return nil, errors.E(op, err)
+	}
+	return ge, nil
+}
+
+// CountGroups returns the number of groups that exist.
+func (j *groupManager) CountGroups(ctx context.Context, user *openfga.User) (int, error) {
+	const op = errors.Op("jimm.CountGroups")
+
+	if !user.JimmAdmin {
+		return 0, errors.E(op, errors.CodeUnauthorized, "unauthorized")
+	}
+	count, err := j.store.CountGroups(ctx)
+	if err != nil {
+		return 0, errors.E(op, err)
+	}
+	return count, nil
+}
+
+// getGroup returns a group based on the provided UUID or name.
+func (j *groupManager) getGroup(ctx context.Context, user *openfga.User, group *dbmodel.GroupEntry) (*dbmodel.GroupEntry, error) {
+	const op = errors.Op("jimm.getGroup")
+
+	if !user.JimmAdmin {
+		return nil, errors.E(op, errors.CodeUnauthorized, "unauthorized")
+	}
+	if err := j.store.GetGroup(ctx, group); err != nil {
+		return nil, errors.E(op, err)
+	}
+	return group, nil
+}
+
+// GetGroupByUUID returns a group based on the provided UUID.
+func (j *groupManager) GetGroupByUUID(ctx context.Context, user *openfga.User, uuid string) (*dbmodel.GroupEntry, error) {
+	return j.getGroup(ctx, user, &dbmodel.GroupEntry{UUID: uuid})
+}
+
+// GetGroupByName returns a group based on the provided name.
+func (j *groupManager) GetGroupByName(ctx context.Context, user *openfga.User, name string) (*dbmodel.GroupEntry, error) {
+	return j.getGroup(ctx, user, &dbmodel.GroupEntry{Name: name})
+}
+
+// RenameGroup renames a group in JIMM's DB.
+func (j *groupManager) RenameGroup(ctx context.Context, user *openfga.User, oldName, newName string) error {
+	const op = errors.Op("jimm.GetGroupManager().RenameGroup")
+
+	if !user.JimmAdmin {
+		return errors.E(op, errors.CodeUnauthorized, "unauthorized")
+	}
+
+	group := &dbmodel.GroupEntry{
+		Name: oldName,
+	}
+	err := j.store.GetGroup(ctx, group)
```

Comment:
hmm.. slightly worried that this is not done in a transaction. while i don't expect us to have much load on the group manger for this to become a problem i still think it should be done in a transaction.